File: packages/sdk-python/tracelane/instrumentations/langgraph.py
"""LangGraph instrumentation for Tracelane.

Wraps LangGraph CompiledGraph.invoke and CompiledGraph.ainvoke to emit OTel
spans for agent graph execution. Captures graph name, step count, and
execution metadata. The predictive layer uses these spans to detect
stuck-loop patterns (PP-PR4).

Example::

    from langgraph.graph import StateGraph, END
    from tracelane.instrumentations.langgraph import instrument_langgraph

    workflow = StateGraph(MyState)
    # ... add nodes, edges
    app = workflow.compile()
    instrument_langgraph(app)
    result = app.invoke({"messages": [...]})
    # Span emitted for the full graph execution
"""

from __future__ import annotations

from typing import Any

from opentelemetry import trace
from opentelemetry.trace import SpanKind, StatusCode

_tracer = trace.get_tracer("tracelane.langgraph", "0.1.0")


def instrument_langgraph(graph: Any) -> None:
    """Wrap a compiled LangGraph graph to emit OTel spans.

    Instruments both synchronous invoke() and asynchronous ainvoke().
    The graph's name (if set) appears as langgraph.graph_name in the span.

    Args:
        graph: A langgraph.graph.CompiledGraph instance.

    Note:
        Step count is extracted from the output if the state has a
        __step__ key (common in multi-turn agent graphs).
        API keys are never captured.
    """
    graph_name = getattr(graph, "name", None) or "unknown"

    _patch_invoke(graph, graph_name)
    _patch_ainvoke(graph, graph_name)
    _patch_stream(graph, graph_name)


def _patch_invoke(graph: Any, graph_name: str) -> None:
    if not hasattr(graph, "invoke"):
        return
    original = graph.invoke

    def _patched_invoke(input: Any, config: Any = None, **kwargs: Any) -> Any:
        with _tracer.start_as_current_span(
            "langgraph.graph.invoke",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "langgraph",
                "langgraph.graph_name": graph_name,
                "langgraph.invocation_type": "invoke",
            },
        ) as span:
            try:
                result = (
                    original(input, config, **kwargs)
                    if config is not None
                    else original(input, **kwargs)
                )
                _record_result(span, result)
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    graph.invoke = _patched_invoke


def _patch_ainvoke(graph: Any, graph_name: str) -> None:
    if not hasattr(graph, "ainvoke"):
        return
    original = graph.ainvoke

    async def _patched_ainvoke(input: Any, config: Any = None, **kwargs: Any) -> Any:
        with _tracer.start_as_current_span(
            "langgraph.graph.ainvoke",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "langgraph",
                "langgraph.graph_name": graph_name,
                "langgraph.invocation_type": "ainvoke",
            },
        ) as span:
            try:
                result = await (
                    original(input, config, **kwargs)
                    if config is not None
                    else original(input, **kwargs)
                )
                _record_result(span, result)
                span.set_status(StatusCode.OK)
                return result
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    graph.ainvoke = _patched_ainvoke


def _patch_stream(graph: Any, graph_name: str) -> None:
    """Wrap stream() to emit a span per chunk for stuck-loop detection."""
    if not hasattr(graph, "stream"):
        return
    original = graph.stream

    def _patched_stream(input: Any, config: Any = None, **kwargs: Any) -> Any:
        with _tracer.start_as_current_span(
            "langgraph.graph.stream",
            kind=SpanKind.CLIENT,
            attributes={
                "gen_ai.provider.name": "langgraph",
                "langgraph.graph_name": graph_name,
                "langgraph.invocation_type": "stream",
            },
        ) as span:
            try:
                gen = (
                    original(input, config, **kwargs)
                    if config is not None
                    else original(input, **kwargs)
                )
                step = 0
                for chunk in gen:
                    step += 1
                    yield chunk
                span.set_attribute("langgraph.step_count", step)
                span.set_status(StatusCode.OK)
            except Exception as exc:
                span.record_exception(exc)
                span.set_status(StatusCode.ERROR, str(exc))
                raise

    graph.stream = _patched_stream


def _record_result(span: Any, result: Any) -> None:
    if isinstance(result, dict):
        step = result.get("__step__")
        if step is not None:
            span.set_attribute("langgraph.step_count", int(step))
        messages = result.get("messages")
        if isinstance(messages, list):
            span.set_attribute("langgraph.messages_count", len(messages))

File: packages/sdk-typescript/src/instrumentations/langgraph.ts
/**
 * LangGraph JS instrumentation for Tracelane.
 *
 * Wraps compiled LangGraph graphs' invoke(), ainvoke(), and stream() methods
 * to emit OTel spans for agent graph execution. Captures graph name, step
 * count, and message count. These spans feed Tracelane's stuck-loop
 * predictor (PP-PR4).
 *
 * @example
 * ```ts
 * import { StateGraph } from "@langchain/langgraph";
 * import { instrumentLangGraph } from "@tracelanedev/sdk/langgraph";
 *
 * const graph = new StateGraph(MyAnnotation).addNode(...).compile();
 * instrumentLangGraph(graph);
 * const result = await graph.invoke({ messages: [...] });
 * ```
 */

import { SpanKind, SpanStatusCode, trace } from "@opentelemetry/api";

const tracer = trace.getTracer("@tracelanedev/sdk-langgraph", "0.1.0");

interface LangGraphLike {
	invoke: (...args: unknown[]) => Promise<unknown>;
	stream?: (...args: unknown[]) => AsyncIterable<unknown>;
	name?: string;
}

/**
 * Instrument a compiled LangGraph graph to emit OTel spans.
 *
 * @param graph - A compiled LangGraph graph instance (CompiledStateGraph or similar)
 */
export function instrumentLangGraph(graph: LangGraphLike): void {
	const graphName = graph.name ?? "unknown";

	_patchInvoke(graph, graphName);
	if (graph.stream) {
		_patchStream(
			graph as LangGraphLike & {
				stream: (...args: unknown[]) => AsyncIterable<unknown>;
			},
			graphName,
		);
	}
}

function _patchInvoke(graph: LangGraphLike, graphName: string): void {
	const originalInvoke = graph.invoke.bind(graph);

	graph.invoke = async (...args: unknown[]) => {
		return tracer.startActiveSpan(
			"langgraph.graph.invoke",
			{
				kind: SpanKind.CLIENT,
				attributes: {
					"gen_ai.provider.name": "langgraph",
					"langgraph.graph_name": graphName,
					"langgraph.invocation_type": "invoke",
				},
			},
			async (span) => {
				try {
					const result = (await originalInvoke(...args)) as Record<
						string,
						unknown
					>;
					_recordResult(span, result);
					span.setStatus({ code: SpanStatusCode.OK });
					return result;
				} catch (e) {
					span.recordException(e as Error);
					span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
					throw e;
				} finally {
					span.end();
				}
			},
		);
	};
}

function _patchStream(
	graph: LangGraphLike & { stream: NonNullable<LangGraphLike["stream"]> },
	graphName: string,
): void {
	const originalStream = graph.stream.bind(graph);

	graph.stream = async function* (...args: unknown[]) {
		const span = tracer.startSpan("langgraph.graph.stream", {
			kind: SpanKind.CLIENT,
			attributes: {
				"gen_ai.provider.name": "langgraph",
				"langgraph.graph_name": graphName,
				"langgraph.invocation_type": "stream",
			},
		});
		let stepCount = 0;
		try {
			for await (const chunk of originalStream(...args)) {
				stepCount++;
				yield chunk;
			}
			span.setAttribute("langgraph.step_count", stepCount);
			span.setStatus({ code: SpanStatusCode.OK });
		} catch (e) {
			span.recordException(e as Error);
			span.setStatus({ code: SpanStatusCode.ERROR, message: String(e) });
			throw e;
		} finally {
			span.end();
		}
	};
}

function _recordResult(
	span: ReturnType<typeof tracer.startSpan>,
	result: Record<string, unknown>,
): void {
	const step = result.__step__;
	if (typeof step === "number") {
		span.setAttribute("langgraph.step_count", step);
	}
	const messages = result.messages;
	if (Array.isArray(messages)) {
		span.setAttribute("langgraph.messages_count", messages.length);
	}
}
